### DISCUSSION TITLE
Slug semantics

### DISCUSSION THREAD

**Discussion by srid** [srid/emanote#42] [Ideas]
Note files get a default URL route right now that is based entirely on their filepath (sans extension). This leads to ugly URLs like `/Foo/Bar%20Qux` (for `/Foo/Bar Qux.md`).

Let's call this default behaviour the "direct" slug semantic. 

- `direct` (default; filepath sans ext)

What other slug semantics are possible? The more pretty ones include,

- `baseonly` (neuron-style)
  - /Foo/Bar => /Bar
- `baseonly-slugify` (neuron-style, but lowercase and whitespace replaced with hyphens)
  - /Foo/Bar Qux => /bar-qux
- `slugify` (like above, but includes fullpath)
  - /Foo/Bar Qux => /foo/bar-qux

The above can be specified at per-route (per-folder) level using YAML configuration.

In addition, we may allow the user to specify a custom slug on a per-Markdown file basis in the frontmatter,

```yaml
url:
  slug: foo
```

This creates `foo.html`. It would create `foo.html` in the containing directory (thus, includes full path) unless that routes's merged YAML config uses one of the "baseonly" semantics (which discards directories in URLs). 

### Progress

- [x] Add support for `slug` in YAML frontmatter, behaving exactly like the feature in neuron: https://github.com/srid/emanote/pull/51
    - Advanced capabilities, as proposed above, maybe added later depending on actual need.
- [x] Allow `slug` to contain `/` (ie. full relative path)
- [ ] Path-based semantics. Worth it? Or not?
    - Worth it, because URLs like this https://tdm.srid.ca/Books/Digital%20Minimalism/Part%201%20-%20Foundations/2%20-%20Digital%20Minimalism are rather ugly, and specifying a (consistent) slug for each .md file is laborious.

**Comment by hi-learn-triz-and-usit** [comment]
> What other slug semantics are possible

- Space conversion behavior and capitalization may be configurable as a options rather than enums (maybe I want lowercase with underscores, or keep case but use hyphens).
- What exactly to do with unicode file names
- Some kind of collision detection and warning
- Some kind of persistent note id (so that if you rename notes, it doesn't break bookmarks)
- Explicit paths for slugs (e.g. `slug: /foo` will create `foo.html` in the root regardless)
- Arbitrary functions of the note in whatever language you like, in the most general case

I think these are roughly in order of usefulness (that is, arbitrary functions are clearly overkill).
